mypy/nodes.py

```diff
@@ -2735,6 +2738,9 @@ def __init__(self, target: 'mypy.types.Type', fullname: str, line: int, column:
         self.alias_tvars = alias_tvars
         self.no_args = no_args
         self.normalized = normalized
+        self.assuming = []  # type: List[Tuple[TypeAliasType, TypeAliasType]]
```

Comment:
Document these?

Is the reason why these "algorithmic" data structures need to be stored on the object themselves (instead of some external map, I guess) because it would require a context being threaded through the whole of the typeops tangle, basically?

Reply:
> Document these?

Yes.
> Is the reason why these "algorithmic" data structures need to be stored on the object themselves (instead of some external map, I guess) because it would require a context being threaded through the whole of the typeops tangle, basically?

Yes, the canonic implementation of `is_subtype()` takes three arguments: `left`, `right`, and `assumptions`, but last time I tried to do this, it required adding `assumptions` to like two dozen functions, and consequently every call site. This is also the reason why we have `TypeInfo.assuming` and friends.